src/update.ts

```diff
@@ -1,149 +1,215 @@
 /* eslint-disable unicorn/prefer-module */
 import color from '@oclif/color'
-import {Config, CliUx} from '@oclif/core'
-import {IManifest} from 'oclif'
+import {Config, CliUx, Interfaces} from '@oclif/core'
 
-import * as spawn from 'cross-spawn'
 import * as fs from 'fs-extra'
 import HTTP from 'http-call'
-import * as _ from 'lodash'
 import * as path from 'path'
+import throttle from 'lodash.throttle'
+import fileSize from 'filesize'
 
 import {extract} from './tar'
-import {ls, wait} from './util'
-
-export interface UpdateCliOptions {
-  channel?: string;
-  autoUpdate: boolean;
-  fromLocal: boolean;
-  config: Config;
-  exit: any;
-  getPinToVersion: () => Promise<string>;
-}
+import {ls, rm, wait} from './util'
 
-export default class UpdateCli {
-  private channel!: string
+const filesize = (n: number): string => {
+  const [num, suffix] = fileSize(n, {output: 'array'})
+  return Number.parseFloat(num).toFixed(1) + ` ${suffix}`
+}
 
-  private currentVersion?: string
+export namespace Updater {
+  export type Options = {
+    channel?: string | undefined;
+    autoUpdate: boolean;
+    version?: string | undefined
+    hard: boolean;
+  }
 
-  private updatedVersion!: string
+  export type VersionIndex = Record<string, string>
+}
 
+export class Updater {
   private readonly clientRoot: string
-
   private readonly clientBin: string
 
-  constructor(private options: UpdateCliOptions) {
-    this.clientRoot = this.options.config.scopedEnvVar('OCLIF_CLIENT_HOME') || path.join(this.options.config.dataDir, 'client')
-    this.clientBin = path.join(this.clientRoot, 'bin', this.options.config.windows ? `${this.options.config.bin}.cmd` : this.options.config.bin)
+  constructor(private config: Config) {
+    this.clientRoot = config.scopedEnvVar('OCLIF_CLIENT_HOME') || path.join(config.dataDir, 'client')
+    this.clientBin = path.join(this.clientRoot, 'bin', config.windows ? `${config.bin}.cmd` : config.bin)
   }
 
-  async runUpdate(): Promise<void> {
-    if (this.options.autoUpdate) await this.debounce()
+  public async runUpdate(options: Updater.Options): Promise<void> {
+    const {autoUpdate, version, hard} = options
+    if (autoUpdate) await this.debounce()
 
-    this.channel = this.options.channel || await this.determineChannel()
+    if (hard) {
```

Comment:
@mdonnalley handling parameter hard here seems premature. Both branches starting at line 53 below call `skipUpdate` which terminates the update command. Would this leave the installation in an unknown state?